src/java/org/apache/cassandra/service/StorageProxy.java

```diff
@@ -1484,7 +1485,11 @@ public static AbstractWriteResponseHandler<IMutation> getWriteResponseHandler(IM
 
         ReplicaPlan.ForTokenWrite replicaPlan = ReplicaPlans.forWrite(keyspace, consistencyLevel, tk, ReplicaPlans.writeNormal);
         AbstractReplicationStrategy rs = replicaPlan.replicationStrategy();
-        return rs.getWriteResponseHandler(replicaPlan, callback, writeType, queryStartNanoTime);
+        AbstractWriteResponseHandler<IMutation> responseHandler = rs.getWriteResponseHandler(replicaPlan, callback, writeType, queryStartNanoTime);
+        if (callback instanceof CounterMutationCallback) {
+            ((CounterMutationCallback) callback).setReplicaCount(replicaPlan.contacts().size());
```

Comment:
because we need access to the `AbstractWriteResponseHandler` itself which by the time `StorageProxy#applyCounterMutationOnLeader` returns it, the callback could have been completed already.

It boils down to the callback being constructed constructed in the verb handler which chronologically happens before the response handler is constructed in `StorageProxy#getWriteResponseHandler` 

Alternative to access the response handler safely in the verb handler is locking + keeping in memory callback -> handler mapping (yikes!) or pass the handler more generally to the callback which requires bigger refactoring (e.g. make callback `Consumer` instead of `Runnable` and more generally pass the response handler instance to it like so `callback.accept(this)`

(other suggestions?)

Reply:
Makes sense, let's stick with the current approach for now.